// typedeck/build/main/models/chipCollection/chipCollection.model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const stringifyComparer_model_1 = require("../../common/stringifyComparer.model");
const chip_service_1 = require("../../services/chip.service");
/**
 * Basic class to represent a grouping of IChips.
 *
 * Supports activities like: getting chips,
 * adding chips, taking specific chip, taking
 * chips by value, getting value of chips.
 */
class ChipCollection {
    constructor(chips = []) {
        this.chips = chips;
        this.objectComparer = new stringifyComparer_model_1.StringifyComparer();
        this.chipService = new chip_service_1.ChipService();
    }
    addChip(chip) {
        return this.addChips([chip]);
    }
    addChips(chips) {
        this.getChips().unshift(...chips);
        return this;
    }
    getChips() {
        return this.chips;
    }
    setChips(chips) {
        this.chips = chips;
        return this;
    }
    getChipCount() {
        return this.getChips().length;
    }
    isEmpty() {
        return this.getChipCount() === 0;
    }
    removeChips(chips) {
        if (chips.length === 0) {
            return this;
        }
        this.chipService.removeChipsFromStack(this.getChips(), chips);
        return this;
    }
    /**
     * Condense chips into fewer but larger denominations.
     * Uses the first chip's class to create new colored chips with.
     */
    colorUp() {
        if (this.getChipCount() === 0) {
            return this;
        }
        const newChips = this.chipService.colorUp(this.getChips(), this.getChips()[0].constructor);
        this.setChips(newChips);
        return this;
    }
    /**
     * Returns first identified index position of chip in
     * collection.
     * @param chip Object to search for in collection
     */
    indexOfChip(chip) {
        for (let i = 0; i < this.getChipCount(); i++) {
            const loopCard = this.getChips()[i];
            if (this.objectComparer.areEquivalent(chip, loopCard)) {
                return i;
            }
        }
        return -1;
    }
    /**
     * Return value of Chips currently in ChipCollection (via `IChipCollection.getChips()`).
     * Optionally, can pass in `IChip[]` to evaluate a manually passed in value.
     * @param chips Optional, value of passed in chips. If no chips passed in,
     * it will evaluate value of current ChipCollection.
     */
    getValue(chips) {
        if (chips === undefined) {
            chips = this.getChips();
        }
        return this.chipService.valueOfChips(chips);
    }
    /**
     * Will pull `IChip`s from `IChipCollection` to satisfy
     * the `amount` requested.
     * @param amount Value to take from IChipCollection
     */
    takeValue(amount) {
        const currentValue = this.getValue();
        if (amount > currentValue) {
            throw new Error(`Not enough chips (${currentValue}) to satisfy requested amount ${amount}`);
        }
        return this.chipService.makeChange(this, amount, this.getChips()[0].constructor);
    }
}
exports.ChipCollection = ChipCollection;
