// typedeck/build/module/models/cardCollection/deck.model.js
import { PlayingCard } from '../card/playingCard.model';
import { CardPile } from './cardPile.model';
import { Hand } from './hand.model';
/**
 * Deck represents the source of cards in play. As such,
 * a Deck is merely a pile of cards that are drawn from
 * throughout the duration of the ongoing game.
 */
export class Deck extends CardPile {
    constructor(cards = []) {
        super(cards);
        this.name = 'Deck';
    }
    static BuildFrom(cards = []) {
        return new Deck(cards);
    }
    static Build(suits, cardNames, extraCards = []) {
        let cards = [];
        suits.forEach((suit) => cardNames.forEach((cardName) => cards.push(new PlayingCard(cardName, suit))));
        if (extraCards && extraCards.length > 0) {
            cards = cards.concat(extraCards);
        }
        return new Deck(cards);
    }
    createHand(options) {
        const hand = new Hand();
        this.deal(hand, options.size, false);
        return hand;
    }
    deal(hand, /** Amount of cards to deal into Hand */ size, /** Deal cards to top of hand */ frontOfHand = false) {
        if (frontOfHand) {
            hand.addCards(this.takeCards(size));
        }
        else {
            hand.addCardsToBottom(this.takeCards(size));
        }
        return this;
    }
}
