// typedeck/build/main/models/chip/chip.model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const chipColor_model_1 = require("./chipColor.model");
const chipColorType_model_1 = require("./chipColorType.model");
/**
 * Represents a generic chip - typically used
 * in Poker style games.
 */
class Chip {
    constructor(color, colorType = chipColorType_model_1.ChipColorType.Standard, overrideValue = 0) {
        this.color = color;
        this.colorType = colorType;
        this.overrideValue = overrideValue;
        this.valueMap = new Map();
    }
    toString() {
        return `${chipColor_model_1.ChipColor[this.color]}`;
    }
    getValue(_color = this.color) {
        if (this.overrideValue !== 0) {
            return this.overrideValue;
        }
        if (this.valueMap.has(_color)) {
            return this.valueMap.get(_color);
        }
        else {
            throw new Error(
            // tslint:disable-next-line:max-line-length
            `Unable to determine value of ${chipColor_model_1.ChipColor[_color]} Chip for ${chipColorType_model_1.ChipColorType[this.colorType]}`);
        }
    }
    setValue(amount) {
        this.overrideValue = amount;
        return this;
    }
    setColorValue(color, amount) {
        this.valueMap.set(color, amount);
        return this;
    }
    getIndex() {
        return this.toString();
    }
}
exports.Chip = Chip;
