// typedeck/build/main/services/chip.service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const chipColor_model_1 = require("../models/chip/chipColor.model");
const standardChip_model_1 = require("../models/chip/standardChip.model");
const iterableExtensions_model_1 = require("../common/iterableExtensions.model");
class ChipService {
    /**
     * The service should accept an `IChipCollection` and
     * a requested specific amount of chip value, and be able to
     * "break" or "make change" to swap out chips to meet that
     * exact need. These chips are removed from the incoming
     * `chipCollection` and the return value are chips that
     * match the `needValue` requested.
     * @param chipCollection Chips to create `needValue` from
     * @param needValue Amount requested from chips
     * @param chipType Class of Chip to return
     */
    makeChange(chipCollection, needValue, chipType = standardChip_model_1.StandardChip, colorUp = false) {
        const currentValue = chipCollection.getValue();
        if (needValue > currentValue) {
            throw new Error(`Not enough chips (${currentValue}) to satisfy requested amount ${needValue}`);
        }
        else if (needValue <= 0) {
            throw new Error(`makeChange requires a positive Chip amount needed`);
        }
        if (colorUp && chipCollection.getChipCount() > 1) {
            const coloredUp = this.sortByValue(this.colorUp(chipCollection.getChips(), chipType));
            chipCollection.setChips(coloredUp);
        }
        // set aside high chips
        let chips = this.sortByValue(chipCollection.getChips());
        let highChips = [];
        if (chipCollection.getChipCount() > 1) {
            let n = chipCollection.getChipCount() - 1;
            while (n > 0) {
                if (this.valueOfChips(chips.slice(0, n)) > needValue) {
                    highChips.push(...chips.splice(n, 1));
                }
                else {
                    break;
                }
                n--;
            }
        }
        let matchedCombination = this.hasCombinationOfAmount(needValue, chips);
        if (matchedCombination.length > 0) {
            chipCollection.removeChips(matchedCombination);
            return [...matchedCombination];
        }
        const breakChip = this.getNextChipToBreak(chips, needValue);
        chipCollection.removeChips([breakChip]);
        const newChips = this.createChips(breakChip.getValue(), false, chipType);
        chipCollection.addChips([...newChips]);
        return this.makeChange(chipCollection, needValue, chipType, false);
    }
    sortByValue(chips) {
        return Array.from(chips)
            .sort((a, b) => a.getValue() - b.getValue());
    }
    createChips(amount, canBeSingleChip = true, chipType = standardChip_model_1.StandardChip) {
        if (amount <= 0) {
            return [];
        }
        let sampleChip = new chipType(chipColor_model_1.ChipColor.White);
        let sortedChips = Array.from(sampleChip.valueMap.entries())
            .sort((a, b) => {
            return a[1] - b[1];
        });
        const chipsThatCanFulfillValue = sortedChips
            .filter((combo) => combo[1] <= amount).length;
        if (chipsThatCanFulfillValue === 1) {
            canBeSingleChip = true;
        }
        let availableChips = sortedChips
            .filter((combo) => canBeSingleChip ? combo[1] <= amount
            : combo[1] < amount)
            .map((entry) => new chipType(entry[0]));
        if (availableChips.length < 1) {
            throw new Error(`Incompatible Chip class to fulfill a value of '${amount}'`);
        }
        let index = availableChips.length - 1;
        const createdChips = [];
        while (amount >= availableChips[0].getValue()) {
            if (amount >= availableChips[index].getValue()) {
                amount -= availableChips[index].getValue();
                createdChips.push(availableChips[index]);
            }
            else {
                index--;
            }
        }
        return createdChips;
    }
    valueOfChips(chips) {
        if (chips.length === 0) {
            return 0;
        }
        return chips.reduce((a, b) => a + b.getValue(), 0);
    }
    colorUp(chips, chipType = standardChip_model_1.StandardChip) {
        const chipsValue = this.valueOfChips(chips);
        const canBeSingleChip = true;
        return this.createChips(chipsValue, canBeSingleChip, chipType);
    }
    hasCombinationOfAmount(amount, chips) {
        const iteratedChips = this.sortByValue(chips);
        let size = chips.length;
        let foundCombination = [];
        while (size > 0) {
            for (let combination of iterableExtensions_model_1.IterableExtensions.Combinations(iteratedChips, size)) {
                if (this.valueOfChips([...combination]) === amount) {
                    foundCombination = [...combination];
                    break; // close iterator
                }
            }
            if (foundCombination.length > 0) {
                return foundCombination;
            }
            size--;
        }
        return [];
    }
    removeChipsFromStack(chips, removeChips) {
        removeChips.forEach((chip) => {
            for (let i = 0; i < chips.length; i++) {
                if (chips[i].getIndex() === chip.getIndex()) {
                    chips.splice(i, 1);
                    break;
                }
            }
        });
        return chips;
    }
    getNextChipToBreak(chips, needValue) {
        const orderedChips = this.sortByValue(chips);
        const reverseOrderedChips = [...orderedChips].reverse();
        const pulledChips = [];
        // find first largest chip at or under value
        let i = 0;
        let runningTotal = 0;
        while (i < reverseOrderedChips.length - 1) {
            const currentChip = reverseOrderedChips[i];
            const addedChipValue = runningTotal + currentChip.getValue();
            if (addedChipValue <= needValue) {
                runningTotal = addedChipValue;
                pulledChips.push(currentChip);
            }
            i++;
        }
        if (runningTotal > 0) {
            let remainingChips = this.removeChipsFromStack([...orderedChips], pulledChips);
            i = 0;
            while (i < remainingChips.length - 1) {
                const currentChip = remainingChips[i];
                const addedChipValue = runningTotal + currentChip.getValue();
                if (addedChipValue <= needValue) {
                    runningTotal = addedChipValue;
                }
                else {
                    // we found our breakchip at remainingChips[i]
                    break;
                }
                i++;
            }
            return remainingChips[i];
        }
        else {
            return reverseOrderedChips[i];
        }
    }
}
exports.ChipService = ChipService;
