// typedeck/build/main/models/cardCollection/cardPile.model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cardCollection_model_1 = require("./cardCollection.model");
/**
 * Extending the basic CardCollection, this class supports
 * dealing with cards in an ordered pile. As such, interacting
 * with the pile from the bottom is supported by using this
 * class.
 */
class CardPile extends cardCollection_model_1.CardCollection {
    constructor(cards = []) {
        super(cards); // CardCollection.constructor
        this.name = 'Pile';
    }
    addCardsToBottom(cards) {
        this.setCards(this.getCards().concat(cards));
    }
    takeCardFromBottom() {
        if (!this.isEmpty()) {
            return this.getCards().pop();
        }
        throw new Error('No cards remaining in pile');
    }
    takeCardsFromBottom(amount) {
        let pulledCards = [];
        while (!this.isEmpty() && pulledCards.length < amount) {
            pulledCards.push(this.getCards().pop());
        }
        return pulledCards;
    }
}
exports.CardPile = CardPile;
