// typedeck/build/main/models/poker/pokerHandResult.model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const pokerHandType_model_1 = require("./pokerHandType.model");
const aceHighRankSet_model_1 = require("../card/aceHighRankSet.model");
class PokerHandResult {
    constructor(cards = [], value = 0, cardsUsed = [], rankSet = new aceHighRankSet_model_1.AceHighRankSet()) {
        /**
         * Comparable value of current hand
         * to rank above or below another
         * `PokerHandResult`.
         */
        this.value = 0;
        /**
         * All cards used to determine
         * result.
         */
        this.cards = [];
        /**
         * Cards in result.
         */
        this.cardsUsed = [];
        this.cards = cards;
        this.value = value;
        this.cardsUsed = cardsUsed;
        this.rankSet = rankSet;
    }
    /**
     * Cards that were used when scoring the hand
     * but had no impact on the score itself. Used to
     * help determine a winner in event of a tied score value.
     */
    get kickers() {
        const cardsNotUsedInResult = this.cards.filter((c) => this.cardsUsed.map(u => u.getIndex()).indexOf(c.getIndex()) === -1)
            .sort((a, b) => this.rankSet.getRankValue(b) - this.rankSet.getRankValue(a));
        return [...cardsNotUsedInResult];
    }
    /**
     * Unique list of `CardName` objects that had a play
     * in `PokerHandResult.value`.
     *
     * If you want every card that played a part
     * in the value, see the `PokerHandResult.cardsUsed` property.
     */
    get scoringHandCardNames() {
        const sortedCardsUsed = this.cardsUsed.sort((a, b) => this.rankSet.getRankValue(b) - this.rankSet.getRankValue(a));
        const uniqueCardNames = new Set(sortedCardsUsed.map(c => c.cardName));
        return [...uniqueCardNames];
    }
    setHandType(type) {
        this.handType = type;
        return this;
    }
    toString() {
        if (this.handType === undefined) {
            return '';
        }
        return pokerHandType_model_1.PokerHandType[this.handType]
            // Look for long acronyms and filter out the last letter
            .replace(/([A-Z]+)([A-Z][a-z])/g, ' $1 $2')
            // Look for lower-case letters followed by upper-case letters
            .replace(/([a-z\d])([A-Z])/g, '$1 $2')
            // Look for lower-case letters followed by numbers
            .replace(/([a-zA-Z])(\d)/g, '$1 $2')
            .replace(/^./, function (str) { return str.toUpperCase(); })
            // Remove any white space left around the word
            .trim();
    }
}
exports.PokerHandResult = PokerHandResult;
