// typedeck/build/module/models/chip/chip.model.js
import { ChipColor } from './chipColor.model';
import { ChipColorType } from './chipColorType.model';
/**
 * Represents a generic chip - typically used
 * in Poker style games.
 */
export class Chip {
    constructor(color, colorType = ChipColorType.Standard, overrideValue = 0) {
        this.color = color;
        this.colorType = colorType;
        this.overrideValue = overrideValue;
        this.valueMap = new Map();
    }
    toString() {
        return `${ChipColor[this.color]}`;
    }
    getValue(_color = this.color) {
        if (this.overrideValue !== 0) {
            return this.overrideValue;
        }
        if (this.valueMap.has(_color)) {
            return this.valueMap.get(_color);
        }
        else {
            throw new Error(
            // tslint:disable-next-line:max-line-length
            `Unable to determine value of ${ChipColor[_color]} Chip for ${ChipColorType[this.colorType]}`);
        }
    }
    setValue(amount) {
        this.overrideValue = amount;
        return this;
    }
    setColorValue(color, amount) {
        this.valueMap.set(color, amount);
        return this;
    }
    getIndex() {
        return this.toString();
    }
}
